locator/tablets.cc

```diff
@@ -0,0 +1,125 @@
+#include "replica/database.hh"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+namespace locator {
+
+seastar::logger tablet_logger("tablets");
+
+std::ostream& operator<<(std::ostream& out, const tablet_replica& r) {
+    return out << r.host << ":" << r.shard;
+}
+
+const tablet_map& tablet_metadata::get_tablet_map(table_id id) const {
+    try {
+        return _tablets.at(id);
+    } catch (const std::out_of_range&) {
+        throw std::runtime_error(format("Tablet map not found for table {}", id));
+    }
+}
+
+tablet_map& tablet_metadata::get_tablet_map(table_id id) {
+    return const_cast<tablet_map&>(
+        const_cast<const tablet_metadata*>(this)->get_tablet_map(id));
+}
+
+void tablet_metadata::set_tablet_map(table_id id, tablet_map map) {
+    _tablets[id] = std::move(map);
+}
+
+future<> tablet_metadata::clear_gently() {
+    for (auto&& [id, map] : _tablets) {
+        co_await map.clear_gently();
+    }
+    co_return;
+}
+
+tablet_map::tablet_map()
+        : tablet_map(1)
+{ }
+
+tablet_map::tablet_map(size_t tablet_count)
+        : _log2_tablets(log2ceil(tablet_count)) {
+    tablet_count = 1ul << _log2_tablets;
+    _tablets.resize(tablet_count);
+}
+
+const tablet_info& tablet_map::get_tablet_info(tablet_id id) const {
+    return _tablets[id];
+}
+
+tablet_id tablet_map::get_tablet_id(token t) const {
+    if (t.is_minimum()) {
+        return 0;
+    }
+    if (t.is_maximum()) {
+        return tablet_count() - 1;
+    }
+    return dht::compaction_group_of(_log2_tablets, t);
+}
+
+dht::token tablet_map::get_last_token(tablet_id id) const {
+    return dht::last_token_of_compaction_group(_log2_tablets, id);
+}
+
+dht::token tablet_map::get_first_token(tablet_id id) const {
+    if (id == 0) {
+        // dht::token::normalize() does not allow std::numeric_limits<int64_t>::min()
+        return dht::token(dht::token_kind::key, std::numeric_limits<int64_t>::min() + 1);
+    } else {
+        return dht::next_token(get_last_token(id - 1));
+    }
+}
+
+dht::token_range tablet_map::get_token_range(tablet_id id) const {
+    if (id == 0) {
+        return dht::token_range::make({dht::minimum_token(), false}, {get_last_token(id), true});
+    } else {
+        return dht::token_range::make({get_last_token(id - 1), false}, {get_last_token(id), true});
+    }
+}
+
+void tablet_map::set_tablet(tablet_id id, tablet_info info) {
+    _tablets[id] = std::move(info);
+}
+
+void tablet_map::set_tablet_transition_info(tablet_id id, tablet_transition_info info) {
+    _transitions[id] = std::move(info);
+}
+
+future<> tablet_map::clear_gently() {
+    for (auto&& info : _tablets) {
+        info.replicas.clear();
+        co_await coroutine::maybe_yield();
+    }
+    _tablets.clear();
+    co_return;
+}
+
+const tablet_transition_info* tablet_map::get_tablet_transition_info(tablet_id id) const {
+    auto i = _transitions.find(id);
+    if (i == _transitions.end()) {
+        return nullptr;
```

Comment:
why does `get_tablet_info` return a reference and throws if `tablet_id` isn't found
but here we return a possibly-null pointer?

Reply:
Because transition info is optional, while tablet info is not.

Reply:
get_tablet_info() throws if there is an out of range access. This could throw in that case as well, but it doesn't seem necessary, it fits with the semantics to treat this as a missing transition.

Reply:
tablet_id has a scope of a tablet map, and tablet map contains info about all tablets in the scope (whole ring). So calling get_tablet_info() with an id which is not present does not mean that you're trying to access a tablet for which the map doesn't contain a mapping, it means the id is wrong and meaningless (it's always a logic error).

Reply:
this merits some good ol'e doc comments